src/hotspot/share/prims/vectorSupport.cpp

```diff
@@ -83,126 +83,87 @@ jint VectorSupport::klass2length(InstanceKlass* ik) {
   return vlen;
 }
 
-void VectorSupport::init_vector_array(typeArrayOop arr, BasicType elem_bt, int num_elem, address value_addr) {
-  int elem_size = type2aelembytes(elem_bt);
-  for (int i = 0; i < num_elem; i++) {
+void VectorSupport::init_payload_element(typeArrayOop arr, bool is_mask, BasicType elem_bt, int index, address addr) {
+  if (is_mask) {
     switch (elem_bt) {
-      case T_BYTE: {
-        jbyte elem_value = *(jbyte*) (value_addr + i * elem_size);
-        arr->byte_at_put(i, elem_value);
-        break;
-      }
-      case T_SHORT: {
-        jshort elem_value = *(jshort*) (value_addr + i * elem_size);
-        arr->short_at_put(i, elem_value);
-        break;
-      }
-      case T_INT: {
-        jint elem_value = *(jint*) (value_addr + i * elem_size);
-        arr->int_at_put(i, elem_value);
-        break;
-      }
-      case T_LONG: {
-        jlong elem_value = *(jlong*) (value_addr + i * elem_size);
-        arr->long_at_put(i, elem_value);
-        break;
-      }
-      case T_FLOAT: {
-        jfloat elem_value = *(jfloat*) (value_addr + i * elem_size);
-        arr->float_at_put(i, elem_value);
-        break;
-      }
-      case T_DOUBLE: {
-        jdouble elem_value = *(jdouble*) (value_addr + i * elem_size);
-        arr->double_at_put(i, elem_value);
-        break;
-      }
-      default:
-        fatal("unsupported: %s", type2name(elem_bt));
+      case T_BYTE:   arr->bool_at_put(index,  (*(jbyte*)addr) != 0); break;
+      case T_SHORT:  arr->bool_at_put(index, (*(jshort*)addr) != 0); break;
+      case T_INT:    // fall-through
+      case T_FLOAT:  arr->bool_at_put(index,   (*(jint*)addr) != 0); break;
+      case T_LONG:   // fall-through
+      case T_DOUBLE: arr->bool_at_put(index,  (*(jlong*)addr) != 0); break;
```

Comment:
Why you push `bool` value for everything?!!!

Reply:
Okay. After reading code more I understand that array is boolean type for case when `is_mask` is `true`. So changes are correct.

Reply:
May be you need separate method to avoid this confusion. The code is separate anyway.